### PR TITLE
Shim implementation for master

### PR DESCRIPTION
**Pull request by amshinde**

### CODE REVIEW

File: shim/shim.c [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+handle_proxy_ctl(struct cc_shim *shim)
+{
+	char buf[LINE_MAX];
+	ssize_t ret;
+
+	if (! shim) {
+		return;
+	}
+
+	ret = read(shim->proxy_sock_fd, buf, LINE_MAX);
```
Comment (jodh-intel) [inline shim/shim.c]: Maybe `LINE_MAX-1` and initialise `buf` to all zeros for safety: see comment below.

File: shim/utils.c [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+#include <string.h>
+#include <errno.h>
+#include <signal.h>
+
+#include "log.h"
+#include "utils.h"
+
+/* Signals that should be forwarded by the shim.
+ * Commented out signals are handled by the default signal handler
+ * as it is appropriate for those signals to be handle by the default handler.
```
Comment (jodh-intel) [inline shim/utils.c]: Typo: `handled`.

File: shim/utils.c [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	//SIGSYS,            /* Bad system call */
+	SIGUNUSED,
+	0,
+};
+
+/*!
+ * Set file descriptor as non-blocking
+ *
+ * \param fd File descriptor to set as non-blocking
+ */
```
Comment (jodh-intel) [inline shim/utils.c]: Missing `\return` comment.

File: shim/utils.c [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        buf[0] = (uint8_t)(val >> 24);
+        buf[1] = (uint8_t)(val >> 16);
+        buf[2] = (uint8_t)(val >> 8);
+        buf[3] = (uint8_t)val;
+}
+
+/*!
+ * Convert the value stored in buffer to little endian
+ *
+ * \param buf Buffer storing the big endian value
```
Comment (jodh-intel) [inline shim/utils.c]: Missing `\return` comment.

File: shim/README.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# `cc-shim`
+
+`cc-shim` is a process spawned by the runtime per container workload. The runtime 
+provides the pid of the cc-shim process to containerd-shim on OCI create command.
+
+Usage:
+	cc-shim --container-id $(container_id) --proxy-sock-fd $(proxy_socket_fd) --proxy-io-fd $(io-fd) --seq-no $(io-seq-no) --err-seq-no $(err-seq-no)
```
Comment (jodh-intel) [inline shim/README.md]: Maybe fold this long line with explicit `\`'s to show the line has been broken.

File: shim/shim.c [hunk 6]
```diff
@@ -0,0 +1,654 @@
...
+/*!
+ * Construct message in the proxy ctl rpc protocol format
+ * Proxy control message format:
+ *
+ * | length  | Reserved| Data(Request/Response   |
+ * | . . . . | . . . . | . . . . . . . . . . . . |
+ * 0         4         8                         length
+ *
+ * \param json Json Payload
+ * \param[out] len Length of the message
```
Comment (jodh-intel) [inline shim/shim.c]: Missing `\return` comment.

File: shim/shim.c [hunk 7]
```diff
@@ -0,0 +1,654 @@
...
+		shim_warning("EOF received on proxy ctl socket. Proxy has exited\n");
+		exit(EXIT_FAILURE);
+	}
+
+	//TODO: Parse the json and log error responses explicitly
+	shim_debug("Proxy response:%s\n", buf + PROXY_CTL_HEADER_SIZE);
+}
+
+long long
+parse_numeric_option(char *input) {
```
Comment (jodh-intel) [inline shim/shim.c]: Missing `\return` comment. Might also be worth noting explicitly (`\note`) that the function can only handle positive values (well, all legitimate values apart from `-1` :-)

**Comment by dlespiau** [comment]
As said earlier, don't seen many problems with merging this and add fixes on top, so:

lgtm

**Comment by jcvenegas** [comment]
lgtm, pullapprove did not take review "lgtm" comment

**Comment by grahamwhaley** [comment]
@jcvenegas I suspect pullapprove nullified it after there was a commit edit :-)

**Comment by chavafg** [comment]
qa-failed

**Comment by chavafg** [comment]
This was make-distcheck again, moving to pass

**Comment by chavafg** [comment]
qa-passed

**Comment by amshinde** [comment]
@jodh-intel  The commit (853aa74e55) I pushed yesterday addresses all your previous review comments.

**Pull request merged**